Doc drop in 3rd party modules

aciba90 opened the pull request:
<!--
Thank you for submitting a PR to cloud-init!

To ease the process of reviewing your PR, do make sure to complete the following checklist **before** submitting a pull request.

- [x] I have signed the CLA: https://ubuntu.com/legal/contributors
- [x] I have added my Github username to ``tools/.github-cla-signers``
- [x] I have included a comprehensive commit message using the guide below
- [ ] I have added unit tests to cover the new behavior under ``tests/unittests/``
  - Test files should map to source files i.e. a source file ``cloudinit/example.py`` should be tested by ``tests/unittests/test_example.py``
  - Run unit tests with ``tox -e py3``
- [x] I have kept the change small, avoiding unnecessary whitespace or non-functional changes.
- [x] I have added a reference to issues that this PR relates to in the PR message (Refs GH-1234, Fixes GH-1234)
- [x] I have updated the documentation with the changed behavior.
  - If the change doesn't change the user interface and is trivial, this step may be skipped.
  - Cloud-config documentation is generated from the jsonschema.
  - Generate docs with ``tox -e docs``.
-->

## Proposed Commit Message
<!-- Include a proposed commit message because PRs are squash merged
by default.

See https://www.conventionalcommits.org/en/v1.0.0/#specification
for our commit message convention.

If the change is related to a particular cloud or particular distro,
please include the "optional scope" in the summary line. E.g.,
feat(ec2): Add support for foo to the baz

Types used by this project:
feat, fix, docs, ci, test, refactor, chore
-->

```
doc: improve drop-in custom modules

Add group of pages for drop-in custom modules and
restructure exitent doc under it.

Add doc for custom datasources and config modules.
    
SC-1836
Fixes GH-4649
```

## Additional Context
<!-- If relevant -->

## Test Steps
<!-- Please include any steps necessary to verify (and reproduce if
this is a bug fix) this change on a live deployed system,
including any necessary configuration files, user-data,
setup, and teardown. Scripts used may be attached directly to this PR. -->


## Merge type

- [x] Squash merge using "Proposed Commit Message"
- [ ] Rebase and merge unique commits. Requires commit messages per-commit each referencing the pull request number (#<PR_NUM>)

blackboxsw commented on doc/rtd/reference/cli.rst:
```diff
@@ -83,30 +83,8 @@ re-run all stages as it did on first boot.
-      sudo rm -rf /var/lib/installer_imgs/
-      sudo rm -rf /var/log/installer/
-
+   What `clean` cleans can be supplemented / customized. See:
```
```suggestion
   What operations `clean` performs can be supplemented / customized. See:
```

TheRealFalcon replied:
Heh, I find the suggested change a bit confusing too. My bikeshed is `The operations performed by `clean` can be ...`

aciba90 replied:
Applied, thanks.

blackboxsw commented on doc/rtd/explanation/format.rst:
```diff
@@ -172,44 +172,8 @@ using a MIME archive.
-   :linenos:
-
-Also, `this blog post`_ offers another example for more advanced usage.
+This is a `part-handler`. It serves for implementing custom formats,
+see :ref:`custom_formats`.
```
It feels a bit strange to obscure this text from this main format page because it also happens to be  custom formatting/extensions because we lose visibility to how a `text/part-handler` could be declared and provided as a user-data format. I think it makes this element a bit harder to discover for user-data types and I'd like to leave the bulk of this content here. Can leave this content in format.rst but cross link or a reference from the generated custom_modules.rst file, but leave this content homed in format.rst?

TheRealFalcon replied:
I also think we should have a larger explanation here, but I'm ok keeping this as-is for now because of the bigger overhaul of this page in my PR.

blackboxsw replied:
works for me. We can await that followup tomorrow which adds more structure to each section and may provide a simple separation between the two references to this material. Ideally I want to be sure that the user-data formats page at least has enough information to represent how the part-handler is declared as a user-data format type. Where the rest of the context/details live is up to your discretion once the second pass on this content provides a pass.

aciba90 replied:
https://github.com/canonical/cloud-init/pull/5548#issuecomment-2252327342

blackboxsw commented on doc/rtd/reference/custom_modules/custom_formats.rst:
```diff
@@ -0,0 +1,51 @@
+.. _custom_formats:
```
Wondering if there is some way to leave this content discoverable in format.rst and just provide a link in the custom_modules.rst that is created. I'd like to make sure minimally the `#part-handler` or `text/part-handler` mime part declaration type is searchable from the user-data format page as all user-data type declarations should probably be seen from that page.

aciba90 replied:
https://github.com/canonical/cloud-init/pull/5548#issuecomment-2252327342

blackboxsw commented on doc/rtd/reference/custom_modules/custom_formats.rst:
```diff
@@ -0,0 +1,51 @@
+**************
+
+One can define custom data formats by presenting a `#part-handler`
+:ref:`user-data format<user_data_formats>`
```
The link doesn't hold a lot of value to jump back to user_data_formats because it just points to our formats page which now redacts all the part-handler specific content.

blackboxsw replied:
Note that there will be collisions with James' PR here too https://github.com/canonical/cloud-init/pull/5551/files

aciba90 replied:
https://github.com/canonical/cloud-init/pull/5548#issuecomment-2252327342

TheRealFalcon commented on doc/rtd/reference/custom_modules/custom_configuration_module.rst:
```diff
@@ -0,0 +1,18 @@
+#. :ref:`Implement a config module<module_creation>` in a Python file with its
+   name starting with ``cc_``.
+
+#. Place the file where the rest of config modules are
```
```suggestion
#. Place the file where the rest of config modules are located.
```

aciba90 replied:
Applied.

TheRealFalcon commented on doc/rtd/reference/custom_modules/custom_cleaners.rst:
```diff
@@ -0,0 +1,28 @@
+
+The :command:`clean` operation is typically performed by image creators
+when preparing a golden image for clone and redeployment. The clean command
+removes any cloud-init semaphores, allowing cloud-init to treat the next
```
I know you didn't write it, but can we change it to `removes any cloud-init stored state` or something similar? Semaphores is an implementation detail that most users won't know about.

aciba90 replied:
Done.

TheRealFalcon commented on doc/rtd/reference/custom_modules/custom_cleaners.rst:
```diff
@@ -0,0 +1,28 @@
+when preparing a golden image for clone and redeployment. The clean command
+removes any cloud-init semaphores, allowing cloud-init to treat the next
+boot of this image as the "first boot". When the image is next booted
+cloud-init will performing all initial configuration based on any valid
```
I'm not sure this sentence is needed at all, but `performing` should be `perform`.

aciba90 replied:
Removed.

TheRealFalcon commented on doc/rtd/reference/custom_modules/custom_cleaners.rst:
```diff
@@ -0,0 +1,28 @@
+datasource meta-data and user-data.
+
+Any executable scripts in this subdirectory will be invoked in lexicographical
+order with run-parts when running the :command:`clean` command.
```
What's run-parts and why is it relevant?

```suggestion
order when running the :command:`clean` command.
```

aciba90 replied:
Removed.

TheRealFalcon commented on doc/rtd/reference/custom_modules/custom_mergers.rst:
```diff
@@ -0,0 +1,66 @@
+Custom Mergers
+**************
+
+TODO adapt and backreference
```
Is this TODO for this PR or for later? If for later, let's remove it before merging.

aciba90 replied:
Removed.

TheRealFalcon commented on doc/rtd/reference/custom_modules/custom_mergers.rst:
```diff
@@ -0,0 +1,66 @@
+
+Because the default
+:ref:`merging<merging_user_data>`
+algorithms and stragies may not always be desired,
```
```suggestion
algorithms and strategies may not always be desired,
```

aciba90 replied:
This was removed.

TheRealFalcon commented on doc/rtd/reference/custom_modules/custom_mergers.rst:
```diff
@@ -0,0 +1,66 @@
+Because the default
+:ref:`merging<merging_user_data>`
+algorithms and stragies may not always be desired,
+the concept of customised merging was introduced through `merge classes`.
```
Nit: now that we've switched back to American English
```suggestion
the concept of customized merging was introduced through `merge classes`.
```

aciba90 replied:
I've removed this phrase in favor of the existent overview that was down.

TheRealFalcon commented on doc/rtd/reference/custom_modules/custom_mergers.rst:
```diff
@@ -0,0 +1,66 @@
+                   merged[k] = v
+           return merged
+
+As you can see, there is an ``_on_dict`` method here that will be given a
```
```suggestion
There is an ``_on_dict`` method here that will be given a
```

aciba90 replied:
Applied.

TheRealFalcon commented on doc/rtd/reference/custom_modules/custom_mergers.rst:
```diff
@@ -0,0 +1,66 @@
+in the :file:`mergers/__init__.py` file (see ``LookupMerger`` and
+``UnknownMerger``).
+
+So, following the typical ``cloud-init`` approach of allowing source code to
```
This sentence seems like a 2nd overview. I think we can drop it entirely, or find a way to move bits to the first line of this section.

aciba90 replied:
Moved up.

TheRealFalcon commented on doc/rtd/reference/custom_modules/custom_mergers.rst:
```diff
@@ -0,0 +1,66 @@
+.. _custom_mergers:
```
This section doesn't address how to use or where to place a custom merger. Can merge syntax be specified in user data for a custom merger like it can for a built in merger?

I know you didn't write this page and are just moving the old sections here, so I don't think that these issues need to be addressed in this PR, but we should probably file a follow-up ticket to address them.

aciba90 replied:
https://github.com/canonical/cloud-init/issues/5560

TheRealFalcon reviewed (commented):
I left some inline comments. 

Each new page is currently fairly small. Do we foresee them growing much? If not, I do wonder if we could collapse all of the new pages into a single "custom modules" page. @s-makin or @holmanb (since you initially filed the issue), any strong opinions?

holmanb commented on doc/rtd/reference/custom_modules.rst:
```diff
@@ -0,0 +1,17 @@
+Custom Modules
+**************
+
+Our reference section contains support information for ``cloud-init``.
```
Looks like this was copied from `doc/rtd/reference/index.rst`. Could we instead write something that describes something about the sub-pages? You can probably grab some text from https://github.com/canonical/cloud-init/issues/4649, but at least some description of what these sections have in common and why a user might turn to customizing cloud-init's behavior would be desirable.

aciba90 replied:
Fixed.

holmanb commented on doc/rtd/reference/custom_modules/custom_cleaners.rst:
```diff
@@ -0,0 +1,28 @@
+Any executable scripts in this subdirectory will be invoked in lexicographical
+order with run-parts when running the :command:`clean` command.
+
+Typical format of such scripts would be a ##-<some-app> like the following:
```
Could we turn this into its own section?

aciba90 replied:
Done.

holmanb commented on doc/rtd/reference/custom_modules/custom_configuration_module.rst:
```diff
@@ -0,0 +1,18 @@
+
+#. Extend the :ref:`base-configuration's <base_config_module_keys>`
+   ``cloud_init_modules``, ``cloud_config_modules`` or ``cloud_final_modules``
+   to let the config module run on one of those stages.
```
I was about to leave a comment about "don't forget to pass a configuration for your custom module to your instance", but that made me realize something else:

I don't think we have a story around jsonschema loading 3rd party configuration modules, do we? That is, 3rd parties would have to manually modify our schema in order to add 3rd party validation. Think that we should document this shortcoming and/or file a bug for it?

aciba90 replied:
https://github.com/canonical/cloud-init/issues/5558

aciba90 replied:
I have added a warning explaining that and the issue.

holmanb commented on doc/rtd/reference/custom_modules/custom_configuration_module.rst:
```diff
@@ -0,0 +1,18 @@
+Custom 3rd-party out-of-tree configuration modules can be added to cloud-init
+by:
+
+#. :ref:`Implement a config module<module_creation>` in a Python file with its
```
I'm wondering if the referenced page wouldn't make more sense here than in the dev docs. What do you think? There may be some overlap in the intent, but since the intent is that users should be able to do this I think that having it under the "custom_modules" umbrella makes more sense. I think that as far as contributing goes, there are probably more useful things for new contributors to be thinking about than writing custom configuration modules for niche purposes.

aciba90 replied:
Yes, I think it is okay either way because:

That content is in the gray are where the persona is an advanced user that is consuming cloud-init as a user but is giving input to cloud-init that requires develoment knowledge about cloud-init's internal.

The same question applies for the documentation on how to create a new datasource and for https://github.com/canonical/cloud-init/issues/5559.

holmanb reviewed (commented):
Thanks for tackling this @aciba90! This is looking great so far. I think we can add some more content to help fill out these pages some more. See my comments below.

> Each new page is currently fairly small. Do we foresee them growing much? If not, I do wonder if we could collapse all of the new pages into a single "custom modules" page. @s-makin or @holmanb (since you initially filed the issue), any strong opinions?

I'd like to see some more content on these pages before we land it, but I really like the structure and organization that this brings to the topics. I'll do a more thorough review tomorrow with some ideas (running out of time now), but as long as we get things properly cross-linked from the appropriate other pages to here, I think that having separate smaller pages that have single focus is a better user experience than large pages that group various things by related topics.

aciba90 commented on doc/rtd/reference/custom_modules/custom_datasource.rst:
```diff
@@ -0,0 +1,19 @@
+#. Place that file in as a single Python module or package in folder included
+   in ``$PYTHONPATH``.
+
+#. Extend the :ref:`base-configuration<base_config_reference>`'s
```
Removed the first one.

aciba90 commented:
Thanks for the reviews, much appreciated! 

As https://github.com/canonical/cloud-init/pull/5551/ is comming and per your recommendations, I have reverted `doc/rtd/explanation/format.rst`, removed 
`doc/rtd/reference/custom_modules/custom_formats.rst` and linked from `custom_modules.rst` to `doc/rtd/explanation/format.rst#part-handler`.

Let me know if something can be further improved. I am happy merging conflicts and / or minimizing the changes surface.

TheRealFalcon reviewed (approved):
LGTM!

blackboxsw commented on doc/rtd/reference/custom_modules/custom_cleaners.rst:
```diff
@@ -0,0 +1,25 @@
+.. _custom_cleaners:
+
+Custom Cleaners
```
Let's make this a bit more descriptive:
```suggestion
.. _custom_clean_scripts:

Custom Clean Scripts
```

blackboxsw commented on doc/rtd/development/datasource_creation.rst:
```diff
@@ -170,6 +170,8 @@ Datasources included in upstream cloud-init benefit from ongoing maintenance,
 development team.
 
+If this is not possible, one can add
+:ref:`custom out-of-tree datasources to cloud-init<custom_datasource>`.
```
nit: pull cloud-init out of this link text (aligns w/ how we setup the link for custom modules in the other file)

blackboxsw commented on doc/rtd/development/module_creation.rst:
```diff
@@ -163,6 +163,17 @@ in the correct location based on dependencies. If your module has no particular
+development team.
+
+If this is not possible, one can add
+:ref:`custom out-of-tree config module<custom_configuration_module>`
```
```suggestion
:ref:`custom out-of-tree config modules<custom_configuration_module>`
```

blackboxsw commented on doc/rtd/reference/cli.rst:
```diff
@@ -83,30 +83,8 @@ re-run all stages as it did on first boot.
-      sudo rm -rf /var/log/installer/
-
+   The operations performed by `clean` can be supplemented / customized. See:
+   :ref:`custom_cleaners`.
```
```suggestion
   :ref:`custom_clean_scripts`.
```

blackboxsw commented on doc/rtd/reference/custom_modules.rst:
```diff
@@ -0,0 +1,24 @@
+.. toctree::
+   :maxdepth: 1
+
+   custom_modules/custom_cleaners.rst
```
```suggestion
   custom_modules/custom_clean_scripts.rst
```

blackboxsw commented on doc/rtd/reference/custom_modules/custom_cleaners.rst:
```diff
@@ -0,0 +1,25 @@
+.. _custom_cleaners:
+
+Custom Cleaners
+***************
```
Let's make this a bit more descriptive (and probably rename the file too)
```suggestion
.. _custom_cleaner_scripts:

Custom Clean Scripts
******************
```

blackboxsw reviewed (approved):
Thanks @aciba90 I think it's worth pushing forward and tracking the significant iterations as separate issues. 

Other inline nits as single diff (other than 
```diff
diff --git a/doc/rtd/development/datasource_creation.rst b/doc/rtd/development/datasource_creation.rst
index 5f8d2b02d..1b6e525b1 100644
--- a/doc/rtd/development/datasource_creation.rst
+++ b/doc/rtd/development/datasource_creation.rst
@@ -171,7 +171,7 @@ compatibility with the rest of the codebase, and security fixes by the upstream
 development team.
 
 If this is not possible, one can add
-:ref:`custom out-of-tree datasources to cloud-init<custom_datasource>`.
+:ref:`custom out-of-tree datasources<custom_datasource>` to cloud-init.
 
 .. _make-mime: https://cloudinit.readthedocs.io/en/latest/explanation/instancedata.html#storage-locations
 .. _DMI: https://www.dmtf.org/sites/default/files/standards/documents/DSP0005.pdf
diff --git a/doc/rtd/development/module_creation.rst b/doc/rtd/development/module_creation.rst
index f9d6d0364..3e10a1ee0 100644
--- a/doc/rtd/development/module_creation.rst
+++ b/doc/rtd/development/module_creation.rst
@@ -172,7 +172,7 @@ compatibility with the rest of the codebase, and security fixes by the upstream
 development team.
 
 If this is not possible, one can add
-:ref:`custom out-of-tree config module<custom_configuration_module>`
+:ref:`custom out-of-tree config modules<custom_configuration_module>`
 to cloud-init.
 
 .. _MetaSchema: https://github.com/canonical/cloud-init/blob/3bcffacb216d683241cf955e4f7f3e89431c1491/cloudinit/config/schema.py#L58
diff --git a/doc/rtd/reference/cli.rst b/doc/rtd/reference/cli.rst
index ca6d35a7a..eb800b22a 100644
--- a/doc/rtd/reference/cli.rst
+++ b/doc/rtd/reference/cli.rst
@@ -84,7 +84,7 @@ re-run all stages as it did on first boot.
 .. note::
 
    The operations performed by `clean` can be supplemented / customized. See:
-   :ref:`custom_cleaners`.
+   :ref:`custom_clean_scripts`.
 
 .. _cli_collect_logs:
 
diff --git a/doc/rtd/reference/custom_modules.rst b/doc/rtd/reference/custom_modules.rst
index 92c84a339..3145e723b 100644
--- a/doc/rtd/reference/custom_modules.rst
+++ b/doc/rtd/reference/custom_modules.rst
@@ -18,7 +18,7 @@ config via user-data or vendor-data.
 .. toctree::
    :maxdepth: 1
 
-   custom_modules/custom_cleaners.rst
+   custom_modules/custom_clean_scripts.rst
    custom_modules/custom_configuration_module.rst
    custom_modules/custom_datasource.rst
    custom_modules/custom_mergers.rst
diff --git a/doc/rtd/reference/custom_modules/custom_cleaners.rst b/doc/rtd/reference/custom_modules/custom_clean_scripts.rst
similarity index 92%
rename from doc/rtd/reference/custom_modules/custom_cleaners.rst
rename to doc/rtd/reference/custom_modules/custom_clean_scripts.rst
index 17d57d445..955668fb2 100644
--- a/doc/rtd/reference/custom_modules/custom_cleaners.rst
+++ b/doc/rtd/reference/custom_modules/custom_clean_scripts.rst
@@ -1,7 +1,7 @@
-.. _custom_cleaners:
+.. _custom_clean_scripts:
 
-Custom Cleaners
-***************
+Custom Clean Scripts
+********************
 
 Cloud-init provides the directory :file:`/etc/cloud/clean.d/` for third party
 applications which need additional configuration artifact cleanup from
```

holmanb reviewed (approved):
Looks good to me! This is an improvement over what we had before.

@blackboxsw your patch on top of this looks good, if you want to apply now and merge.
